src/java.base/share/classes/java/lang/runtime/ObjectMethods.java

```diff
@@ -251,44 +255,109 @@ private static MethodHandle makeHashCode(Class<?> receiverClass,
-        String formatString = sb.toString();
-        MethodHandle formatter = MethodHandles.insertArguments(STRING_FORMAT, 0, formatString)
-                                              .asCollector(String[].class, getters.size()); // (R*)String
-        if (getters.size() == 0) {
-            // Add back extra R
-            formatter = MethodHandles.dropArguments(formatter, 0, receiverClass);
+
+        boolean firstTime = true;
+        MethodHandle[] mhs;
+        List<List<MethodHandle>> splits;
+        MethodHandle[] toSplit = getters;
+        int namesIndex = 0;
+        do {
+            /* StringConcatFactory::makeConcatWithConstants can only deal with 200 spots, longs and double occupy two
+             * the rest 1 spot, we need to chop the current `getters` into chuncks, it could be that for records with
+             * a lot of components that we need to do a couple of iterations. The main diference between the first
+             * iteration and the rest would be on the recipe
+             */
+            splits = split(toSplit);
+            mhs = new MethodHandle[splits.size()];
+            for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+                String recipe = "";
+                if (firstTime && splitIndex == 0) {
+                    recipe = receiverClass.getSimpleName() + "[";
+                }
+                for (int i = 0; i < splits.get(splitIndex).size(); i++) {
+                    recipe += firstTime ? names.get(namesIndex) + "=" + "\1" : "\1";
+                    if (firstTime && namesIndex != names.size() - 1) {
+                        recipe += ", ";
+                    }
+                    namesIndex++;
+                }
+                if (firstTime && splitIndex == splits.size() - 1) {
+                    recipe += "]";
+                }
+                Class<?>[] concatTypeArgs = new Class<?>[splits.get(splitIndex).size()];
+                // special case: no need to create another getters if there is only one split
+                MethodHandle[] currentSplitGetters = new MethodHandle[splits.get(splitIndex).size()];
+                for (int j = 0; j < splits.get(splitIndex).size(); j++) {
+                    concatTypeArgs[j] = splits.get(splitIndex).get(j).type().returnType();
+                    currentSplitGetters[j] = splits.get(splitIndex).get(j);
+                }
+                MethodType concatMT = MethodType.methodType(String.class, concatTypeArgs);
+                try {
+                    mhs[splitIndex] = StringConcatFactory.makeConcatWithConstants(
+                            lookup, "",
+                            concatMT,
+                            recipe,
+                            new Object[0]
+                    ).getTarget();
+                    mhs[splitIndex] = MethodHandles.filterArguments(mhs[splitIndex], 0, currentSplitGetters);
+                    mhs[splitIndex] = MethodHandles.permuteArguments(
```

Comment:
This is some gnarly logic. Could we add some comments on what permuteArguments with a reorder array of just zeros is doing here?

Reply:
This is not unusual. It spreads a single argument across several "getters". But a comment wouldn't hurt.

Reply:
I haven't seen it used like this before. The duplication behavior is mentioned explicitly with an example in the javadoc, sure, but it still took me a deep reading of the docs to understand what was going on here and what the intent was.